haskell-lsp-types/src/Language/Haskell/LSP/Types/DataTypesJSON.hs

```diff
@@ -2586,6 +2611,37 @@ deriveJSON lspOptions ''RenameParams
 type RenameRequest  = RequestMessage ClientMethod RenameParams WorkspaceEdit
 type RenameResponse = ResponseMessage WorkspaceEdit
 
+-- ---------------------------------------------------------------------
+{-
+Prepare Rename Request
+
+Since version 3.12.0
+
+The prepare rename request is sent from the client to the server to setup
+and test the validity of a rename operation at a given location.
+
+Request:
+
+    method: ‘textDocument/prepareRename’
+    params: TextDocumentPositionParams
+
+Response:
+
+    result: Range | { range: Range, placeholder: string } | null describing
+            the range of the string to rename and optionally a placeholder
+            text of the string content to be renamed. If null is returned
+            then it is deemed that a ‘textDocument/rename’ request is not
+            valid at the given position.
+    error: code and message set in case an exception happens during the
+           prepare rename request.
+
+-}
+
+-- {\"jsonrpc\":\"2.0\",\"id\":1,\"method\":\"textDocument/rename\",\"params\":{\"textDocument\":{\"uri\":\"file:///home/alanz/mysrc/github/alanz/haskell-lsp/src/HieVscode.hs\"},\"position\":{\"line\":37,\"character\":17},\"newName\":\"getArgs'\"}}
+
+type PrepareRenameRequest  = RequestMessage ClientMethod TextDocumentPositionParams Range
+type PrepareRenameResponse = ResponseMessage Range
```

Comment:
Should we handle the case where it can return a range and a placeholder?

Reply:
Yes, I'll add a type for that

Reply:
Just added the type salad for this